### PR TITLE
feat: Configure java connector to check CN instance name. Fixes #1995

### LINKED ISSUE #1995: Check that the server certificate CN matches the instance name
### Feature Description

During the TLS authentication, the server certificate CN field should match the instance name. Add a custom TrustManager that will check the CN field and reject it if the name does not match the instance name.

### PR DESCRIPTION
**Pull request by hessjcg**
This code will make that the connector checks that the server certificate's subject CN field matches the instance name
before allowing the TLS handshake to proceed.

This introduces new classes that are always installed as trust managers on all TLS connections to ensure that the TLS logic correctly checks that the server certificate's CN field matches the instance.

-  `InstanceCheckingTrustManager` - Delegates to default TLS trust manager, then checks the CN field.
-  `InstanceCheckingTrustManagerFactory` and `InstanceCheckingTrustManagerFactorySpi` - Installs
     our custom trust manager into connector TLS sockets.

The `ConscryptWorkaroundTrustManagerFactory` and `ConscryptWorkaroundTrustManagerFactorySpi` classes are no longer necessary. Logic to detect Conscrypt and use the workaround `ConscryptWorkaroundTrustManager` moves into `InstanceCheckingTrustManagerFactorySpi`.

### CODE REVIEW

**Review by enocom — COMMENTED** [review]
Do we have a test that verifies a common name mismatch fails the TLS handshake?

**Comment by hessjcg** [comment]
I added a test for a CN mismatch.

**Comment by jackwotherspoon** [comment]
> @jackwotherspoon We need to manually test this with a domain-scoped instance. Can you do that?

@jonath sure, for manually testing should I just run the integration tests? Will that suffice?

**Comment by hessjcg** [comment]
@jackwotherspoon: Yes, it will be sufficient to run the integration tests at the domain-scoped instance.

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlCoreTestingBase.java [hunk 1]
```diff
@@ -62,6 +62,8 @@ public class CloudSqlCoreTestingBase {
 
   ListenableFuture<KeyPair> clientKeyPair;
 
+  public CloudSqlCoreTestingBase() {}
```
Comment (enocom) [inline core/src/test/java/com/google/cloud/sql/core/CloudSqlCoreTestingBase.java:65]: FWIW In my opinion, this base class deserves to be split up into small composable units. As is, it's an awkward tool.

File: core/src/test/java/com/google/cloud/sql/core/ConnectorTest.java [hunk 2]
```diff
@@ -80,6 +81,32 @@ public void create_throwsErrorForInvalidInstanceName() throws IOException {
...
   }
 
+  @Test
+  public void create_throwsErrorForInvalidTlsCommonNameMismatch()
+      throws IOException, InterruptedException {
+    // The server TLS certificate matches myProject:myRegion:myInstance
+    FakeSslServer sslServer = new FakeSslServer();
+    ConnectionConfig config =
+        new ConnectionConfig.Builder()
+            .withCloudSqlInstance("myProject:myRegion:otherInstance")
```
Comment (enocom) [inline core/src/test/java/com/google/cloud/sql/core/ConnectorTest.java:91]: nit: let's make this blatantly wrong, e.g., "wrongwrongwrong" or something similar that catches the eye.

**Review by enocom — APPROVED** [review]

File: core/src/test/java/com/google/cloud/sql/core/CloudSqlCoreTestingBase.java (hunk 1 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/CloudSqlCoreTestingBase.java:65]: +1. Maybe in a different PR.

File: core/src/test/java/com/google/cloud/sql/core/ConnectorTest.java (hunk 2 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/ConnectorTest.java:91]: Fixed.

**Pull request merged**